src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -63,24 +63,16 @@ internal static class DeserializationHelpers
             if (enableCaseInsensitiveModelBinding && edmProperty == null)
             {
                 // if edmProperty is null, we try a Case-Insensitive match
-                IDictionary<string, string> structuralPropertyNamesMapping = new Dictionary<string, string>();
-                var sProperties = resourceType.StructuralProperties();
-                foreach (var prop in sProperties)
+                var sProperties = resourceType.StructuralProperties().Where(k => k.Name.Equals(property.Name, StringComparison.OrdinalIgnoreCase));
+
+                // We throw an exception when we have two case-insensitive matches and no case sensitive matches.
+                if (sProperties.Count() > 1)
                 {
-                    try
-                    {
-                        structuralPropertyNamesMapping.Add(prop.Name.ToUpperInvariant(), prop.Name);
-                    }
-                    catch (ArgumentException)
-                    {
-                        // Exception when we have two case-insensitive matches and no case sensitive matches.
-                        throw new ODataException(Error.Format(SRResources.CannotDeserializeUnknownProperty, property.Name, resourceType.Definition));
-                    }
+                    throw new ODataException(Error.Format(SRResources.CannotDeserializeUnknownProperty, property.Name, resourceType.Definition));
                 }
 
-                string name;
-                var findPropertyName = structuralPropertyNamesMapping.TryGetValue(property.Name.ToUpperInvariant(), out name) ? name : property.Name;
-                edmProperty = resourceType.FindProperty(findPropertyName);
+                var sProperty = sProperties.First();
```

Comment:
Minor nit: I'd go with .`Single` instead of `.First` since we know it's only supposed to be 1 item at this point
Q: is there any likelihood that `sProperties` might be empty at this point in which case we might want to consider `.[First|Single]OrDefault`? Or do we want it to fail instead of returning null?

Reply:
I have added a condition to ensure `sProperties` is not empty. When `sProperties` is empty, `edmProperty` will remain as `null` and exception will be thrown later in subsequent code.
#Resolved